bootenv/keydata_test.go

```diff
@@ -0,0 +1,457 @@
+	primaryKey := s.newPrimaryKey(c, 32)
+
+	params := &KeyDataScopeParams{
+		PrimaryKey: primaryKey,
+		Role:       data.validRole,
+		KDFAlg:     data.kDFAlg,
+		MDAlg:      data.mDAlg,
+		ModelAlg:   data.modelAlg,
+	}
+
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	err = kds.SetAuthorizedSnapModels(primaryKey, data.role, data.validModels...)
+	if err != nil {
+		return err
+	}
+
+	err = kds.SetAuthorizedBootModes(primaryKey, data.role, data.validModes...)
+	if err != nil {
+		return err
+	}
+
+	SetModel(data.model)
+	SetBootMode(data.bootMode)
+
+	return kds.IsBootEnvironmentAuthorized()
+}
+
+func (s *keyDataPlatformSuite) TestBootEnvAuthValid1(c *C) {
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+	}
+
+	validModes := []string{
+		"modeFoo",
+	}
+
+	s.mockState(c)
+
+	c.Check(s.testBootEnvAuth(c, &testBootEnvAuthData{
+		kDFAlg:      crypto.SHA256,
+		mDAlg:       crypto.SHA256,
+		modelAlg:    crypto.SHA256,
+		validRole:   "test",
+		role:        "test",
+		validModels: validModels,
+		model:       validModels[0],
+		validModes:  validModes,
+		bootMode:    validModes[0],
+	}), IsNil)
+}
+
+func (s *keyDataPlatformSuite) TestBootEnvAuthValid2(c *C) {
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+		s.makeMockModelAssertion(c, "model-b"),
+	}
+
+	validModes := []string{
+		"modeFoo",
+	}
+
+	s.mockState(c)
+
+	c.Check(s.testBootEnvAuth(c, &testBootEnvAuthData{
+		kDFAlg:      crypto.SHA256,
+		mDAlg:       crypto.SHA256,
+		modelAlg:    crypto.SHA256,
+		validRole:   "test",
+		role:        "test",
+		validModels: validModels,
+		model:       validModels[1],
+		validModes:  validModes,
+		bootMode:    validModes[0],
+	}), IsNil)
+}
+
+func (s *keyDataPlatformSuite) TestBootEnvAuthInvalidModel(c *C) {
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+	}
+
+	invalidModel := s.makeMockModelAssertion(c, "model-b")
+
+	validModes := []string{
+		"modeFoo",
+	}
+
+	s.mockState(c)
+
+	c.Check(s.testBootEnvAuth(c, &testBootEnvAuthData{
+		kDFAlg:      crypto.SHA256,
+		mDAlg:       crypto.SHA256,
+		modelAlg:    crypto.SHA256,
+		role:        "test",
+		validRole:   "test",
+		validModels: validModels,
+		model:       invalidModel,
+		validModes:  validModes,
+		bootMode:    validModes[0],
+	}), ErrorMatches, "unauthorized model")
+}
+
+func (s *keyDataPlatformSuite) TestBootEnvAuthInvalidBootMode(c *C) {
+	validModels := []SnapModel{
+		s.makeMockModelAssertion(c, "model-a"),
+	}
+
+	validModes := []string{
+		"modeFoo",
+	}
+
+	invalidBootMode := "modeBar"
+
+	s.mockState(c)
+
+	c.Check(s.testBootEnvAuth(c, &testBootEnvAuthData{
+		kDFAlg:      crypto.SHA256,
+		mDAlg:       crypto.SHA256,
+		modelAlg:    crypto.SHA256,
+		validRole:   "test",
+		role:        "test",
+		validModels: validModels,
+		model:       validModels[0],
+		validModes:  validModes,
+		bootMode:    invalidBootMode,
+	}), ErrorMatches, "unauthorized boot mode")
+}
```

Comment:
Whilst the test coverage here is mostly ok (other than missing the JSON encoding / decoding), it's important to have a test that verifies a specific primary key always derives the same elliptic key - I'll leave it up to you to decide how best to test that, but it will require some hard-coded keys in the test code. There are some examples of this kind of test already (eg, `policyV3SuiteNoTPM.TestDerivePolicyAuthKey*`).

Reply:
Added a test for that too. While adding the tests for marshalling/unmarshalling, I accidentally came across the fact that Go's `ecdsa.GenerateKey` is not deterministic. After adding the test I modified `KeyDataScope.deriveSigner()` to use `ecdsa.GenerateKey()` instead of the internal crypto out of curiosity and my test failed all the times I ran it (I am running 10 iterations). 

Interestingly many of the times it failed even before the test reached my loop, between the 2 derivations that happen in `NewKeyDataScope` itself (one when filling the PublicKey field and one in `authorize()`)

Reply:
Thanks. Yes, go1.20 changed the behaviour of `ecdsa.GenerateKey` to use a non-deterministic method to generate the key, which is a bit of a pain.

The new tests do check that multiple calls to `KeyDataScope.deriveSigner` produce the same key, but there also needs to be a test that verifies it produces the expected key for a fixed input (ie, a test with a hardcoded primary key and expected derived key).

Reply:
I added 2 tests for 2 different fixed keys and another one that tests that different keys are derived from the same primary key but a different role which was also missing.